packages/office-ui-fabric-react/src/components/ComboBox/ComboBox.tsx

```diff
@@ -0,0 +1,1130 @@
+      return currentPendingValue !== '' ? currentPendingValue :
+        (this._indexWithinBounds(currentOptions, index) ? currentOptions[index].text : '');
+    } else {
+
+      // If we are not allowing freeform and have a
+      // valid index that matches the pending value,
+      // we know we will need some version of the pending value
+      if (this._indexWithinBounds(currentOptions, currentPendingValueValidIndex)) {
+
+        // If autoComplete is on, return the
+        // raw pending value, otherwise remember
+        // the matched option's index
+        if (autoComplete) {
+          return currentPendingValue;
+        }
+
+        index = currentPendingValueValidIndex;
+      }
+
+      // If we have a valid index then return the text value of that option,
+      // otherwise return the suggestedDisplayValue
+      return this._indexWithinBounds(currentOptions, index) ? currentOptions[index].text : suggestedDisplayValue;
+    }
+  }
+
+  /**
+   * Is the index within the bounds of the array?
+   * @param options - options to check if the index is valid for
+   * @param index - the index to check
+   * @returns {boolean} - true if the index is valid for the given options, false otherwise
+   */
+  @autobind
+  private _indexWithinBounds(options: IComboBoxOption[], index: number): boolean {
+    return index >= 0 && index < options.length;
+  }
+
+  /**
+   * Handler for typing changes on the input
+   * @param updatedValue - the newly changed value
+   */
+  @autobind
+  private _onInputChange(updatedValue: string) {
+    this.props.allowFreeform ?
+      this._processInputChangeWithFreeform(updatedValue) :
+      this._processInputChangeWithoutFreeform(updatedValue);
+  }
+
+  /**
+   * Process the new input's new value when the comboBox
+   * allows freeform entry
+   * @param updatedValue - the input's newly changed value
+   */
+  @autobind
+  private _processInputChangeWithFreeform(updatedValue: string) {
+    let {
+      currentOptions
+    } = this.state;
+
+    // if the new value is empty, nothing needs to be done
+    if (updatedValue === '') {
+      return;
+    }
+
+    // Remember the original value and then,
+    // make the value lowercase for comparison
+    let originalUpdatedValue: string = updatedValue;
+    updatedValue = updatedValue.toLocaleLowerCase();
+
+    let newSuggestedDisplayValue = '';
+    let newCurrentPendingValueValidIndex = -1;
+
+    // If autoComplete is on, attempt to find a match from the available options
+    if (this.props.autoComplete) {
+
+      // If autoComplete is on, attempt to find a match where the text of an option starts with the updated value
+      let items = currentOptions.map((item, index) => { return { ...item, index }; }).filter((option) => option.itemType !== DropdownMenuItemType.Header && option.itemType !== DropdownMenuItemType.Divider).filter((option) => option.text.toLocaleLowerCase().indexOf(updatedValue) === 0);
```

Comment:
I can imagine the autocomplete algorithm to be configurable as well, but probably not worth adding until needed.